src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java

```
@@ -397,32 +396,45 @@ protected void doExecute(Task task, ActionRequest actionRequest, ActionListener<
      *
      * @param failure  object that may contain exceptions thrown
      * @param detector detector object
-     * @return whether cold start runs
+     * @return exception if we get resource not found exception
      * @throws AnomalyDetectionException List of exceptions we can throw
      *     1. Exception from cold start:
      *       1). InternalFailure due to
      *         a. ElasticsearchTimeoutException thrown by putModelCheckpoint during cold start
      *       2). EndRunException with endNow equal to false
      *         a. training data not available
      *         b. cold start cannot succeed
+     *         c. invalid training data
+     *       3) EndRunException with endNow equal to true
+     *         a. invalid search query
      *     2. LimitExceededException from one of RCF model node when the total size of the models
      *      is more than X% of heap memory.
      *     3. InternalFailure wrapping ElasticsearchTimeoutException inside caused by
      *      RCF/Threshold model node failing to get checkpoint to restore model before timeout.
      */
-    private boolean coldStartIfNoModel(AtomicReference<AnomalyDetectionException> failure, AnomalyDetector detector)
+    private Exception coldStartIfNoModel(AtomicReference<AnomalyDetectionException> failure, AnomalyDetector detector)
         throws AnomalyDetectionException {
         AnomalyDetectionException exp = failure.get();
-        if (exp != null) {
-            if (exp instanceof ResourceNotFoundException) {
-                LOG.info("Trigger cold start for {}", detector.getDetectorId());
-                globalRunner.compute(new ColdStartJob(detector));
-                return true;
-            } else {
-                throw exp;
+        if (exp == null) {
+            return null;
+        }
+
+        if (!(exp instanceof ResourceNotFoundException)) {
+            throw exp;
```

Why not return this exception ?

---

The returned exception is for cold start related failure only.